connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyTransformer.java

```diff
@@ -0,0 +1,108 @@
+    public static final String CLIENT = "client";
+    public static final String CLIENT_URL = "client_url";
+    public static final String CONTEXT = "context";
+    public static final String CUSTOM_DETAILS = "custom_details";
+    public static final String DISPLAY_NAME = "display_name";
+    public static final String ENVIRONMENT_URL = "environment_url";
+    public static final String EVENT_ACTION = "event_action";
+    public static final String EVENT_TYPE = "event_type";
+    public static final String EVENTS = "events";
+    public static final String GROUP = "group";
+    public static final String ORG_ID = "org_id";
+    public static final String SEVERITY = "severity";
+    public static final String SOURCE = "source";
+    public static final String SOURCE_NAMES = "source_names";
+    public static final String SUMMARY = "summary";
+    public static final String TIMESTAMP = "timestamp";
+
+    public static final DateTimeFormatter PD_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+
+    @Override
+    public void process(Exchange exchange) {
+        JsonObject cloudEventData = exchange.getIn().getBody(JsonObject.class);
+
+        JsonObject cloudEventPayload = cloudEventData.getJsonObject(PAYLOAD);
+        JsonObject message = new JsonObject();
+        message.put(EVENT_ACTION, PagerDutyEventAction.TRIGGER);
+        message.mergeIn(getClientLink(cloudEventPayload, cloudEventData.getString(ENVIRONMENT_URL)));
+
+        JsonObject messagePayload = new JsonObject();
+        messagePayload.put(SUMMARY, cloudEventPayload.getString(EVENT_TYPE));
+        messagePayload.put(TIMESTAMP, LocalDateTime.parse(cloudEventPayload.getString(TIMESTAMP)).format(PD_DATE_TIME_FORMATTER));
+        // TODO is it possible for tenant apps to indicate what severity to use?
+        messagePayload.put(SEVERITY, PagerDutySeverity.WARNING);
+        messagePayload.put(SOURCE, cloudEventPayload.getString(APPLICATION));
+        messagePayload.put(GROUP, cloudEventPayload.getString(BUNDLE));
+
+        JsonObject customDetails = new JsonObject();
+        customDetails.put(ACCOUNT_ID, cloudEventPayload.getString(ACCOUNT_ID));
+        customDetails.put(ORG_ID, cloudEventPayload.getString(ORG_ID));
+        customDetails.put(CONTEXT, JsonObject.mapFrom(cloudEventPayload.getJsonObject(CONTEXT)));
+        customDetails.put(SOURCE_NAMES, JsonObject.mapFrom(cloudEventPayload.getJsonObject(SOURCE)));
+        // Keep events, if provided
+        if (cloudEventPayload.containsKey(EVENTS)) {
+            customDetails.put(EVENTS, cloudEventPayload.getJsonArray(EVENTS));
+        }
+
+        messagePayload.put(CUSTOM_DETAILS, customDetails);
+        message.put(PAYLOAD, messagePayload);
+
+        exchange.getIn().setBody(message.encode());
+    }
+
+    /**
+     * Adapted from CamelProcessor template for Teams
+     * <br>
+     * TODO should this be a template in notifications-engine for maintainability?
```

Comment:
[Here is the template](https://github.com/RedHatInsights/notifications-backend/blob/71bc5427d96431aa6928444890dde24c1b6d30a8/engine/src/test/java/com/redhat/cloud/notifications/processors/camel/teams/TeamsProcessorTest.java#L17-L22) I used to get the display text and URL.

Reply:
Here's a template used in prod:
```
{#if data.context.display_name??}
<{data.environment_url}{#if data.bundle == "openshift" and data.application == "advisor"}/openshift/insights/advisor/clusters/{data.context.display_name}{#else}/insights/inventory/{#if data.context.inventory_id??}{data.context.inventory_id}{#else}?hostname_or_id={data.context.display_name}{/if}{/if}|{data.context.display_name}> triggered {data.events.size()} event{#if data.events.size() > 1}s{/if}{#else}
{data.events.size()} event{#if data.events.size() > 1}s{/if} triggered{/if} from {data.source.application.display_name} - {data.source.bundle.display_name}. <{data.environment_url}/{#if data.bundle == "openshift"}openshift/{/if}{#if data.application == "integrations"}settings/{#else}insights/{/if}{data.application}|Open {data.source.application.display_name}>
```
The URL will need more work but that can be done in a later PR.